### ISSUE TITLE
Implement the new Status Notifier spec for the system tray

### ISSUE THREAD

**Issue by paulolieuthier** [lxde/lxqt#359]
We can use KF5's KNotifications's [KStatusNotifierItem](http://api.kde.org/frameworks-api/frameworks5-apidocs/knotifications/html/classKStatusNotifierItem.html) class for that (and we could make use of other bits of KNotifications as well). See Martin's words about that [here](http://blog.martin-graesslin.com/blog/2014/03/system-tray-in-plasma-next/) and [here](http://blog.martin-graesslin.com/blog/2014/06/where-are-my-systray-icons/).

We should maintain a X11 XEmbed version of a systray to support old applications.

**Comment by paulolieuthier** [comment]
@surlykke: https://launchpad.net/sni-qt.

**Comment by surlykke** [comment]
2015-01-12 15:14 GMT+01:00 Paulo Lieuthier <notifications@github.com>:

> @surlykke <https://github.com/surlykke>: https://launchpad.net/sni-qt.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lxde/lxqt/issues/359#issuecomment-69574865>.
>
Ah - nice. Thx.

**Comment by dglent** [comment]
I saw the new systray of kf5 and i don't like it
It has some bugs:
at the menu 'Cities' the submenu is ignored:
http://glenbox.free.fr/files/screenshots/pyqt5_systray.png

Another incombatibility (among others)
The hovered signal is not emitted:
http://glenbox.free.fr/files/screenshots/hovered_not_emitted.png

It works in a 'normal' system tray:
http://glenbox.free.fr/files/screenshots/hovered_emitted.png

And i don't like that it adds an entry 'quit'. 
Why add a menu 'Quit' in an application, leave application's programmers decide about theirs context menu entries ...

**Comment by equeim** [comment]
I wrote simple plugin to show StatusNotifierItems.
https://github.com/equeim/lxqt-panel-plugin-snw
It doesn't require KNotifications, but uses dbusmenu-qt5. It doesn't support Xembed tray.
Also Qt 5.5 will have it own built-in implementation of StatusNotifierItems and dbusmenu.

**Comment by jleclanche** [comment]
Well done. Ill look at it soon.

I didnt know Qt 5.5 would have its own implementation of it. Could be worth just backporting it.

**Comment by equeim** [comment]
>>I didnt know Qt 5.5 would have its own implementation of it. Could be worth just backporting it.

https://github.com/qtproject/qtbase/tree/dev/src/platformsupport/dbusmenu
https://github.com/qtproject/qtbase/tree/dev/src/platformsupport/dbustray

**Comment by surlykke** [comment]
Seems others are doing stuff on this, and since I've not gotten around to doing anything, I'll un-assign my self.

**Comment by paulolieuthier** [comment]
@equeim awesome! I'll try it later.

**Comment by paulolieuthier** [comment]
@equeim:
```sh
➜  build git:(master) cmake ..
-- The C compiler identification is GNU 4.9.2
-- The CXX compiler identification is GNU 4.9.2
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.28") 
-- checking for module 'dbusmenu-qt5'
--   found dbusmenu-qt5, version 0.9.2
-- Configuring done
-- Generating done
-- Build files have been written to: /home/paulo/Workspace/lxqt/lxqt-panel-plugin-snw/build
➜  build git:(master) make
Scanning dependencies of target snw_automoc
[ 16%] Automatic moc for target snw
Generating moc_snwplugin.cpp
/home/paulo/Workspace/lxqt/lxqt-panel-plugin-snw/src/snwplugin.h:24: Error: Undefined interface
AUTOGEN: error: process for /home/paulo/Workspace/lxqt/lxqt-panel-plugin-snw/build/src/moc_snwplugin.cpp failed:
/home/paulo/Workspace/lxqt/lxqt-panel-plugin-snw/src/snwplugin.h:24: Error: Undefined interface

Generating moc_statusnotifierbutton.cpp
Generating moc_statusnotifierwatcher.cpp
Generating moc_statusnotifierwidget.cpp
moc failed...
src/CMakeFiles/snw_automoc.dir/build.make:49: recipe for target 'src/CMakeFiles/snw_automoc' failed
make[2]: *** [src/CMakeFiles/snw_automoc] Error 1
CMakeFiles/Makefile2:111: recipe for target 'src/CMakeFiles/snw_automoc.dir/all' failed
make[1]: *** [src/CMakeFiles/snw_automoc.dir/all] Error 2
Makefile:137: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by equeim** [comment]
Are you sure that lxqt-panel is installed? I've got the same error when removed it. There is no way to check if panel is installed, cmake checks only liblxqt.

**Comment by jleclanche** [comment]
It's probably time to think of a decent architecture for third party plugins. I would like us to move some plugins out of the tree and we keep getting new ones.

**Comment by paulolieuthier** [comment]
@equeim that was needed to build it:
```diff
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 2198ec3..67f1b93 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -13,6 +13,7 @@ set(SOURCES
 )
 
 set(CMAKE_AUTOMOC On)
+include(${QTXDG_USE_FILE})
 
 include_directories(${LXQT_INCLUDE_DIRS} ${QTXDG_INCLUDE_DIRS} ${dbusmenu-qt5_INCLUDE_DIRS})
 add_library(snw SHARED ${HEADERS} ${SOURCES})
diff --git a/src/snwplugin.h b/src/snwplugin.h
index f8b9153..b4bdab7 100644
--- a/src/snwplugin.h
+++ b/src/snwplugin.h
@@ -2,6 +2,7 @@
 #define SNWPLUGIN_H
 
 #include "statusnotifierwidget.h"
+#include <ilxqtpanelplugin.h>
 
 class SNWPlugin : public QObject, public ILxQtPanelPlugin
 {
```
Great work, I'll get in touch soon.

**Comment by paulolieuthier** [comment]
@equim another thing, this line:
```cmake
install(TARGETS snw DESTINATION lib/lxqt-panel)
```
should become something similar to:
```cmake
include(GNUInstallDirs)
install(TARGETS snw DESTINATION ${CMAKE_INSTALL_FULL_LIBDIR}/lxqt-panel)
```

**Comment by paulolieuthier** [comment]
Just to make ir clear, KNotifications [has become tier 3](http://api.kde.org/frameworks-api/frameworks5-apidocs/knotifications/html/knotifications-dependencies.html), which adds significant new dependencies. We should keep looking at @equeim's work.

**Comment by jleclanche** [comment]
I don't think there's any point to use KNotifications anyway.

**Comment by paulolieuthier** [comment]
@jleclanche KStatusNotifierItem, which is what we meant to use in the beginning for adapting to the new spec, is shiped in KNotifications.

**Comment by paulolieuthier** [comment]
Landed in lxde/lxqt-panel@1d6731b62ab32d903622c2f0278c87a3ff6a1d90.
